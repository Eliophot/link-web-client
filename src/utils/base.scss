// Breakpoints
$xsMax: 575px;
$smMin: 576px;
$smMax: 767px;
$mdMin: 768px;
$mdMax: 991px;
$lgMin: 992px;
$lgMax: 1199px;
$xlgMin: 1200px;

// Colors
$mainColor: #4696e5;
$lightHoverColor: #eeeeee;
$lightGrey: #dddddd;
$dangerColor: #dc3545;
$mediumGrey: #dee2e6;

// Misc
$headerHeight: 57px;
