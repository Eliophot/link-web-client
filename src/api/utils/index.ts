import { AxiosError } from 'axios';
import {
  ErrorTypeV2,
  ErrorTypeV3,
  InvalidArgumentError,
  InvalidShortUrlDeletion,
  ProblemDetailsError,
  RegularNotFound,
} from '../types/errors';

const isProblemDetails = (e: unknown): e is ProblemDetailsError =>
  !!e && typeof e === 'object' && Object.keys(e).every((key) => ['type', 'detail', 'title', 'status'].includes(key));

const isAxiosError = (e: unknown): e is AxiosError<ProblemDetailsError> => !!e && typeof e === 'object' && 'response' in e;

export const parseApiError = (e: unknown): ProblemDetailsError | undefined => {
  if (isProblemDetails(e)) {
    return e;
  }

  return (isAxiosError(e) ? e.response?.data : undefined);
};

export const isInvalidArgumentError = (error?: ProblemDetailsError): error is InvalidArgumentError =>
  error?.type === ErrorTypeV2.INVALID_ARGUMENT || error?.type === ErrorTypeV3.INVALID_ARGUMENT;

export const isInvalidDeletionError = (error?: ProblemDetailsError): error is InvalidShortUrlDeletion =>
  error?.type === 'INVALID_SHORTCODE_DELETION'
  || error?.type === ErrorTypeV2.INVALID_SHORT_URL_DELETION
  || error?.type === ErrorTypeV3.INVALID_SHORT_URL_DELETION;

export const isRegularNotFound = (error?: ProblemDetailsError): error is RegularNotFound =>
  (error?.type === ErrorTypeV2.NOT_FOUND || error?.type === ErrorTypeV3.NOT_FOUND) && error?.status === 404;
