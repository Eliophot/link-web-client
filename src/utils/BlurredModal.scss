@import './base';

@media (min-width: $smMin) {
  #root {
    transition: filter 300ms;
  }

  .with-modal #root {
    filter: blur(1.5px);
  }
}
