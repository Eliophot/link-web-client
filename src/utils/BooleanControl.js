import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { v4 as uuid } from 'uuid';

export const basePropTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([ PropTypes.string, PropTypes.node ]),
  className: PropTypes.string,
};

const propTypes = {
  ...basePropTypes,
  type: PropTypes.oneOf([ 'switch', 'checkbox' ]).isRequired,
};

const BooleanControl = ({ checked, onChange, className, children, type }) => {
  const id = uuid();
  const onChecked = (e) => onChange(e.target.checked, e);
  const typeClasses = {
    'custom-switch': type === 'switch',
    'custom-checkbox': type === 'checkbox',
  };

  return (
    <span className={classNames('custom-control', typeClasses, className)} style={{ display: 'inline' }}>
      <input type="checkbox" className="custom-control-input" id={id} checked={checked} onChange={onChecked} />
      <label className="custom-control-label" htmlFor={id}>{children}</label>
    </span>
  );
};

BooleanControl.propTypes = propTypes;

export default BooleanControl;
