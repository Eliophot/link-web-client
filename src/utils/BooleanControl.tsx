import { ChangeEvent, FC } from 'react';
import classNames from 'classnames';
import { v4 as uuid } from 'uuid';
import { identity } from 'ramda';

export interface BooleanControlProps {
  checked?: boolean;
  onChange?: (checked: boolean, e: ChangeEvent<HTMLInputElement>) => void;
  className?: string;
}

interface BooleanControlWithTypeProps extends BooleanControlProps {
  type: 'switch' | 'checkbox';
}

const BooleanControl: FC<BooleanControlWithTypeProps> = (
  { checked = false, onChange = identity, className, children, type },
) => {
  const id = uuid();
  const onChecked = (e: ChangeEvent<HTMLInputElement>) => onChange(e.target.checked, e);
  const typeClasses = {
    'custom-switch': type === 'switch',
    'custom-checkbox': type === 'checkbox',
  };

  return (
    <span className={classNames('custom-control', typeClasses, className)}>
      <input type="checkbox" className="custom-control-input" id={id} checked={checked} onChange={onChecked} />
      <label className="custom-control-label" htmlFor={id}>{children}</label>
    </span>
  );
};

export default BooleanControl;
