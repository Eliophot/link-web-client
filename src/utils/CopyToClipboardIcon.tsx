import { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCopy as copyIcon } from '@fortawesome/free-regular-svg-icons';
import CopyToClipboard from 'react-copy-to-clipboard';
import './CopyToClipboardIcon.scss';

interface CopyToClipboardIconProps {
  text: string;
  onCopy?: (text: string, result: boolean) => void;
}

export const CopyToClipboardIcon: FC<CopyToClipboardIconProps> = ({ text, onCopy }) => (
  <CopyToClipboard text={text} onCopy={onCopy}>
    <FontAwesomeIcon icon={copyIcon} className="ms-2 copy-to-clipboard-icon" />
  </CopyToClipboard>
);
