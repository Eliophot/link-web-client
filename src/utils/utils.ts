import { isEmpty, isNil, pipe, range } from 'ramda';
import type { SyntheticEvent } from 'react';

export const rangeOf = <T>(size: number, mappingFn: (value: number) => T, startAt = 1): T[] =>
  range(startAt, size + 1).map(mappingFn);

export type Empty = null | undefined | '' | never[];

export const hasValue = <T>(value: T | Empty): value is T => !isNil(value) && !isEmpty(value);

export const handleEventPreventingDefault = <T>(handler: () => T) => pipe(
  (e: SyntheticEvent) => e.preventDefault(),
  handler,
);

export type Nullable<T> = {
  [P in keyof T]: T[P] | null
};

type Optional<T> = T | null | undefined;

export type OptionalString = Optional<string>;

export const nonEmptyValueOrNull = <T>(value: T): T | null => (isEmpty(value) ? null : value);

export const capitalize = <T extends string>(value: T): string => `${value.charAt(0).toUpperCase()}${value.slice(1)}`;

export const equals = (value: any) => (otherValue: any) => value === otherValue;

export type BooleanString = 'true' | 'false';

export const parseBooleanToString = (value: boolean): BooleanString => (value ? 'true' : 'false');

export const parseOptionalBooleanToString = (value?: boolean): BooleanString | undefined => (
  value === undefined ? undefined : parseBooleanToString(value)
);
