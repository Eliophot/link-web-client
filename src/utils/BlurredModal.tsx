import { FC } from 'react';
import { ModalProps, Modal } from 'reactstrap'; // eslint-disable-line import/named
import './BlurredModal.scss';

const onEnter = () => document.body.classList.add('with-modal');
const onExit = () => document.body.classList.remove('with-modal');

export const BlurredModal: FC<ModalProps> = ({ children, ...rest }) => (
  <Modal {...rest} modalTransition={{ onEnter, onExit }}>
    {children}
  </Modal>
);
