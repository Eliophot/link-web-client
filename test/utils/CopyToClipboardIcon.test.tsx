import { shallow, ShallowWrapper } from 'enzyme';
import CopyToClipboard from 'react-copy-to-clipboard';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCopy as copyIcon } from '@fortawesome/free-regular-svg-icons';
import { CopyToClipboardIcon } from '../../src/utils/CopyToClipboardIcon';

describe('<CopyToClipboardIcon />', () => {
  let wrapper: ShallowWrapper;
  const onCopy = () => {};

  beforeEach(() => {
    wrapper = shallow(<CopyToClipboardIcon text="foo" onCopy={onCopy} />);
  });
  afterEach(() => wrapper?.unmount());

  test('expected components are wrapped', () => {
    const copyToClipboard = wrapper.find(CopyToClipboard);
    const icon = wrapper.find(FontAwesomeIcon);

    expect(copyToClipboard).toHaveLength(1);
    expect(copyToClipboard.prop('text')).toEqual('foo');
    expect(copyToClipboard.prop('onCopy')).toEqual(onCopy);
    expect(icon).toHaveLength(1);
    expect(icon.prop('icon')).toEqual(copyIcon);
    expect(icon.prop('className')).toEqual('ml-2 copy-to-clipboard-icon');
  });
});
