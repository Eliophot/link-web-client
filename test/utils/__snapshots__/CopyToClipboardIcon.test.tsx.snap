// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CopyToClipboardIcon /> wraps expected components 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-copy ms-2 copy-to-clipboard-icon"
    data-icon="copy"
    data-prefix="far"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M224 0c-35.3 0-64 28.7-64 64V288c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64H224zM64 160c-35.3 0-64 28.7-64 64V448c0 35.3 28.7 64 64 64H288c35.3 0 64-28.7 64-64V384H304v64c0 8.8-7.2 16-16 16H64c-8.8 0-16-7.2-16-16V224c0-8.8 7.2-16 16-16h64V160H64z"
      fill="currentColor"
    />
  </svg>
</div>
`;
